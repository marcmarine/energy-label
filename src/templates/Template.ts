import { QRCodeGenerator } from '../utils'
import type { ArrowData } from './ArrowTemplate'
import { TemplateCommon } from './TemplateCommon'

export abstract class Template<T = ArrowData> {
  protected data?: Partial<T>

  async generate(data: Partial<T>): Promise<string> {
    this.data = data

    const background = this.createBackground()
    const header = await this.createHeader()
    const efficiencyScale = this.createEfficiencyScale()
    const consumption = this.createConsumption()
    const features = this.createFeatures()

    return this.render([background, header, efficiencyScale, consumption, features])
  }

  protected createBackground(): string {
    return ''
  }

  protected async createHeader(): Promise<string> {
    return ''
  }

  protected abstract createEfficiencyScale(): string

  protected createConsumption(): string {
    return ''
  }

  protected createFeatures(): string {
    return ''
  }

  protected abstract getWidth(): number
  protected abstract getHeight(): number

  protected async generateQRCodeDataUrl(registrationNumber: string): Promise<string> {
    const url = `${TemplateCommon.EPREL_BASE_URL}/${registrationNumber}`
    const options = { margin: 0, width: 512 }
    return await QRCodeGenerator.generate(url, options)
  }

  protected render(sections: string[]): string {
    const width = this.getWidth()
    const height = this.getHeight()
    const content = sections.join('')

    return `<svg xmlns="http://www.w3.org/2000/svg" width="${width}" height="${height}" viewBox="0 0 ${width} ${height}" fill="none">${content}</svg>`
  }
}
