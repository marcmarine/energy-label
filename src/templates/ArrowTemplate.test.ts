import { describe, it, expect, beforeEach, vi } from 'vitest'
import { Template } from './Template'
import { ArrowTemplate } from './ArrowTemplate'

describe('Arrow Template', () => {
  it('should generate complete arrow template', async () => {
    const template = new ArrowTemplate()
    const result = await template.generate({})

    expect(result).toMatchSnapshot()
  })
})
