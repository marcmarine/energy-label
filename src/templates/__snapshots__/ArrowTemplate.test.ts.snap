// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Arrow Template > should generate complete arrow template 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="92" height="51" viewBox="0 0 92 51" fill="none"><path fill="#fff" d="M90.74.41H23L.5 25 23 49.71h67.74V.41Z" /><path fill="#00A651" d="M72 .41H23L.5 25 23 49.71h49V.41Z" /><path stroke="black" stroke-linecap="round" stroke-linejoin="round" stroke-width="0.8pt" d="M90.74.41H23L.5 25 23 49.71h67.74V.41Z" /><text font-size="29pt" font-weight="bold" fill="white" font-family="Calibri" text-anchor="middle" stroke="black" stroke-width="0.8"><tspan x="46" y="37">A</tspan></text><path stroke="black" stroke-linecap="round" stroke-linejoin="round" stroke-width="1pt" d="M81.33 30.37V20.01M84.08 22.71l-2.75-2.75-2.74 2.75" /><path fill="#000" d="M86.38 35.86c0 .18-.01.334-.028.464-.012.13-.034.235-.065.315a.305.305 0 0 1-.102.158.21.21 0 0 1-.14.046c-.08 0-.21-.055-.389-.167a5.761 5.761 0 0 0-.714-.361 7.628 7.628 0 0 0-1.067-.362 5.46 5.46 0 0 0-1.429-.167 3.92 3.92 0 0 0-1.577.306 3.35 3.35 0 0 0-1.206.853c-.334.365-.59.807-.77 1.327a5.146 5.146 0 0 0-.269 1.707c0 .687.09 1.293.27 1.818.185.52.441.956.77 1.309.327.352.72.618 1.177.797.464.18.974.27 1.531.27.272 0 .541-.032.807-.093.266-.062.513-.155.742-.279v-2.783h-2.273c-.11 0-.197-.068-.26-.204-.061-.142-.092-.38-.092-.714 0-.174.006-.319.018-.436.019-.118.044-.21.075-.279a.357.357 0 0 1 .111-.157.222.222 0 0 1 .148-.056h4.055c.099 0 .185.018.26.056.08.03.148.08.204.148a.658.658 0 0 1 .139.25c.03.093.046.202.046.325v4.778c0 .186-.034.35-.102.492-.062.136-.195.244-.399.324a7.87 7.87 0 0 1-.742.25 10.19 10.19 0 0 1-1.893.363c-.321.03-.646.046-.974.046-.959 0-1.818-.136-2.579-.408a5.296 5.296 0 0 1-1.92-1.197 5.249 5.249 0 0 1-1.206-1.911c-.279-.755-.418-1.608-.418-2.56 0-.984.149-1.865.445-2.645a5.551 5.551 0 0 1 1.262-1.994 5.43 5.43 0 0 1 1.967-1.262c.773-.29 1.633-.436 2.58-.436.519 0 .992.037 1.418.111.427.068.798.155 1.114.26.321.1.587.207.798.325.21.117.355.22.436.306a.75.75 0 0 1 .185.371c.037.155.056.387.056.696ZM86.88 15.09c.075.223.121.4.14.53.019.13-.006.229-.074.296-.068.062-.189.1-.362.112a7.075 7.075 0 0 1-.714.028c-.316 0-.563-.007-.743-.019a1.902 1.902 0 0 1-.398-.046.39.39 0 0 1-.195-.12.637.637 0 0 1-.093-.196l-.807-2.412h-4.509l-.76 2.347a.74.74 0 0 1-.103.223.407.407 0 0 1-.195.14c-.086.03-.213.052-.38.064-.16.012-.374.019-.64.019a6.16 6.16 0 0 1-.668-.028c-.16-.019-.272-.062-.334-.13-.062-.074-.084-.176-.065-.306.019-.13.065-.303.14-.52l3.7-10.65a.857.857 0 0 1 .13-.25.422.422 0 0 1 .233-.149c.11-.037.262-.059.454-.065.192-.012.445-.019.76-.019.366 0 .656.007.873.019.217.006.383.028.501.065.124.03.21.08.26.148.055.068.102.161.139.279l3.71 10.64Zm-5.51-8.803h-.01l-1.697 5.102h3.405L81.37 6.287Z" /></svg>"`;
